### PR TITLE
Added option to make hunter traps invisible

### LINKED ISSUE #1927: Trap Visibility (Renewal)
### Current Behavior
Traps are visible to everyone.

### Expected Behavior
Official:
Hunter traps are only visible to the caster or his/her party/guildmates/alliance in all maps.
Official Video: https://www.youtube.com/watch?v=lSDfPb905-w&feature=youtu.be
Official Video with party: https://youtu.be/fdWpcQirGuo?t=124
 
I suggest to make a settings like:
// Trap Settings
// 1: Only hunters can see the traps.
// 2: Only the Hunter and his/her party can see the trap. (Official - Renewal)
// 3: All traps are visible to everyone (pre-renewal).
trap_visible: 2

**Branch(es):**
- [x ] master
- [ ] other

**Hercules rev. hash/commit:** 

Git revision src:

[//]: # (Copy the first 3 lines of the login-server, char-server or map-server startup.)
[//]: # ( [Info]: Hercules 64-bit for Mac OS X )
[/x/]: # ( [Info]: Git revision src: 'a5918b329ca0826b04dca32ede783586403f58db' )
[/x/]: # ( [Info]: Git revision scripts: 'a5918b329ca0826b04dca32ede783586403f58db' )

*edit* Only the Hunter and his/her party can see the trap on official.

### LINKED ISSUE #1928: Trap - Detect
Official:
All traps can be detected by Detect skill of Hunters
Official Video: https://www.youtube.com/watch?v=lSDfPb905-w&feature=youtu.be

### PR DESCRIPTION
**Pull request by guilherme-gm**
[//]: # (**********************************)
[//]: # (** Fill in the following fields **)
[//]: # (**********************************)

[//]: # (Note: Lines beginning with syntax such as this one, are comments and will not be visible in your report!)

### Pull Request Prelude

[//]: # (Thank you for working on improving Hercules!)

[//]: # (Please complete these steps and check the following boxes by putting an `x` inside the brackets _before_ filing your Pull Request.)

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR will be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

[//]: # (Describe at length, the changes that this pull request makes.)
This PR adds a new SkillInfo flag: `HiddenTrap`, that's used to make certain traps invisible (according to `trap_options` battle conf).
This makes traps visibility work like in renewal (configurable to also be used like in pre-renewal and some custom formats).
`traps_setting` was deprecated and `trap_options` was created with 2 sub configs:
-  `visibility`: if they should be visible/invisible once set on the ground
    * 0: Traps are always visible to everyone (same as traps_setting = 0), and that's original/pre-renewal configuration
    * 1: traps marked with "HiddenTrap" are invisible in versus maps, but visible in other maps, until it's detected
    * 2: Traps marked with 'Hiddentrap" are invisible in any map, until it's detected

- `display_on_trigger`: if they should become visible once triggered
    * 0: No (except for Ankle Snare, that becomes visible) -- this is how it officially works
    * 1: Yes (set as default -- read below for why)

I set display_on_trigger as yes by default because not displaying them makes it really weird, e.g. for Skid Trap, if you keep it hidden it'll look like you lagged a lot and went back 5 cels, when you actually triggered a trap.

Notes:
- Traps are always visible to players in the same party as the hunter who cast the trap (this is done by the client itself).

**Affected Branches:** 
master

**Issues addressed:**
Fixes #1927 
Fixes #1928 

[//]: # (Issue Tracker Number if any.)

### Known Issues and TODO List

[//]: # (Insert checklist here)
[//]: # (Syntax: - [ ] Checkbox)

[//]: # (**NOTE** Enable the setting "[√] Allow edits from maintainers." when creating your pull request if you have not already enabled it.)

Trap list/settings based on rathena/rathena#986
Special thanks to @kyeme 

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

**Comment by HerculesWSAPI** [comment]
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/2232)

File: src/map/skill.c [hunk 1]
```diff
@@ -12759,6 +12758,11 @@ static int skill_unit_onplace_timer(struct skill_unit *src, struct block_list *b
 			ts->tick += sg->interval*(map->count_oncell(bl->m,bl->x,bl->y,BL_CHAR,0)-1);
 	}
 
+	if(battle_config.trap_visibility != 0 && skill->get_inf2(sg->skill_id) & INF2_HIDDEN_TRAP) {
```
Comment (4144) [inline src/map/skill.c]: better add space after ``if`` and before ``(``

File: src/map/skill.c [hunk 2]
```diff
@@ -17051,6 +17055,13 @@ static struct skill_unit *skill_initunit(struct skill_unit_group *group, int idx
 	su->val1=val1;
 	su->val2 = val2;
 	su->prev = 0;
+	su->visible = true;
+
+	if (skill->get_inf2(group->skill_id) & INF2_HIDDEN_TRAP
+		&& ((battle_config.trap_visibility == 1 && map_flag_vs(group->map)) // invisible in PvP/GvG
+			|| battle_config.trap_visibility == 2 // always invisible
+		))
+	 	su->visible = false;
```
Comment (4144) [inline src/map/skill.c]: for complex if like this, better add ``{}`` before/after code inside if block.
``su->visible = false;`` -->
```
{
    su->visible = false;
}
```

File: src/map/skill.c [hunk 3]
```diff
@@ -11040,9 +11039,9 @@ static int skill_castend_pos2(struct block_list *src, int x, int y, uint16 skill
 			map->foreachinarea(status->change_timer_sub,
 			                   src->m, x-r, y-r, x+r,y+r,BL_CHAR,
 			                   src,NULL,SC_SIGHT,tick);
-			if(battle_config.traps_setting&1)
-			map->foreachinarea(skill_reveal_trap,
-			                   src->m, x-r, y-r, x+r, y+r, BL_SKILL);
+			if(battle_config.trap_visibility != 0)
```
Comment (4144) [inline src/map/skill.c]: here need space and ``{}`` too

File: src/map/skill.c [hunk 4]
```diff
@@ -17051,6 +17057,14 @@ static struct skill_unit *skill_initunit(struct skill_unit_group *group, int idx
 	su->val1=val1;
 	su->val2 = val2;
 	su->prev = 0;
+	su->visible = true;
+
+	if (skill->get_inf2(group->skill_id) & INF2_HIDDEN_TRAP
+		&& ((battle_config.trap_visibility == 1 && map_flag_vs(group->map)) // invisible in PvP/GvG
+			|| battle_config.trap_visibility == 2 // always invisible
+		)) {
+	 	su->visible = false;
```
Comment (4144) [inline src/map/skill.c]: here bit formatting issue ``{`` placed at same column with code inside block.
should be ``)) {`` moved bit left or code in block bit right

**Comment by guilherme-gm** [comment]
@4144 I think this pr is fine, can you review it again? Thanks!

**Pull request merged**